### PR TITLE
add documentation for running wasm workloads

### PR DESCRIPTION
**Pull request by devigned**
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**:
This PR adds documentation for running Wasm / WASI workloads on CAPZ clusters.

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
add documentation for running WebAssembly workloads on CAPZ clusters
```

### CODE REVIEW

**Review by CecileRobertMichon — COMMENTED** [review]
doc looks good on first read but I'm going to try this out and follow the steps for my own learning

File: docs/book/src/topics/wasi.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
+# WebAssembly / WASI Workloads
+
+## Overview
+
+CAPZ enables you to create WebAssembly (Wasm) / WASI pod workloads targeting either [Deislabs Slight](https://github.com/deislabs/spiderlightning) or [Fermyon Spin](https://github.com/fermyon/spin) frameworks for building and running fast, secure microservices with Wasm on Kubernetes (v1.23.16+, v1.24.10+, v1.25.6+, v1.26.1+, and newer Kubernetes versions).
+
+Both of the runtimes (slight and spin) for running Wasm workloads use [Wasmtime](https://wasmtime.dev) embedded in containerd shims via the [deislabs/containerd-wasm-shims](https://github.com/deislabs/containerd-wasm-shims) project which is built upon [containerd/runwasi](https://github.com/containerd/runwasi). These containerd shims enable Kubernetes to run Wasm workloads without needing to embed the Wasm runtime in each OCI image. 
+
+## Slight (SpiderLightning)
+"SpiderLightning defines a set of WebAssembly Interface Types (i.e., WIT) files that abstract distributed application capabilities, such as state management, pub/sub, event driven programming, and more."
```
Comment (Mossaka) [inline docs/book/src/topics/wasi.md:10]: May need to leave a more descriptive way to introduce `slight`. 

"slight is a open source wasmtime-based runtime that provide cloud capabilities to Wasm microservices. These capabilites include keyvalue, pub/sub, and much more. "

**Review by Mossaka — APPROVED** [review]

File: docs/book/src/topics/wasi.md (hunk 1 above)
Comment (devigned) [reply docs/book/src/topics/wasi.md]: Great suggestion.

File: docs/book/src/topics/wasi.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
+# WebAssembly / WASI Workloads
+
+## Overview
+
+CAPZ enables you to create WebAssembly (Wasm) / WASI pod workloads targeting either [Deislabs Slight](https://github.com/deislabs/spiderlightning) or [Fermyon Spin](https://github.com/fermyon/spin) frameworks for building and running fast, secure microservices with Wasm on Kubernetes (v1.23.16+, v1.24.10+, v1.25.6+, v1.26.1+, and newer Kubernetes versions).
```
Comment (mboersma) [inline docs/book/src/topics/wasi.md:5]:
```suggestion
CAPZ enables you to create WebAssembly (Wasm) / WASI pod workloads targeting either [Deislabs Slight](https://github.com/deislabs/spiderlightning) or [Fermyon Spin](https://github.com/fermyon/spin) frameworks for building and running fast, secure microservices on Kubernetes (v1.23.16+, v1.24.10+, v1.25.6+, v1.26.1+, and newer Kubernetes versions).
```

File: docs/book/src/topics/wasi.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+Both of the runtimes (slight and spin) for running Wasm workloads use [Wasmtime](https://wasmtime.dev) embedded in containerd shims via the [deislabs/containerd-wasm-shims](https://github.com/deislabs/containerd-wasm-shims) project which is built upon [containerd/runwasi](https://github.com/containerd/runwasi). These containerd shims enable Kubernetes to run Wasm workloads without needing to embed the Wasm runtime in each OCI image. 
+
+## Slight (SpiderLightning)
+Slight (or Spiderlightning) is an open source wasmtime-based runtime that provide cloud capabilities to Wasm microservices. These capabilities include key/value, pub/sub, and much more.
```
Comment (mboersma) [inline docs/book/src/topics/wasi.md:10]:
```suggestion
Slight (or Spiderlightning) is an open source wasmtime-based runtime that provides cloud capabilities to Wasm microservices. These capabilities include key/value, pub/sub, and much more.
```

File: docs/book/src/topics/wasi.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+handler: "slight"
+---
+apiVersion: node.k8s.io/v1
+kind: RuntimeClass
+metadata:
+  name: "wasmtime-spin-v1"
+handler: "spin"
+```
+
+The preceding runtime classes will register a runtime class for slight and spin, which will direct containerd to use the spin or slight shim when a pod workload is scheduled onto a cluster node.
```
Comment (mboersma) [inline docs/book/src/topics/wasi.md:33]:
```suggestion
The preceding YAML document will register a runtime class for slight and spin, which will direct containerd to use the spin or slight shim when a pod workload is scheduled onto a cluster node.
```

File: docs/book/src/topics/wasi.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  type: LoadBalancer
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 80
+  selector:
+    app: wasm-spin
+```
+
+The preceding deployment and service will create a load balanced "hello world" service with 3 Spin microservices. Note the `runtimeClassName` applied to the Deployment, `wasmtime-spin-v1`, which informs containerd on the cluster node to run the workload with the spin shim.
```
Comment (mboersma) [inline docs/book/src/topics/wasi.md:81]:
```suggestion
The preceding deployment and service will create a load-balanced "hello world" service with 3 Spin microservices. Note the `runtimeClassName` applied to the Deployment, `wasmtime-spin-v1`, which informs containerd on the cluster node to run the workload with the spin shim.
```

File: docs/book/src/topics/wasi.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+kind: RuntimeClass
+metadata:
+  name: "wasmtime-spin-v1"
+handler: "spin"
+scheduling:
+  nodeSelector:
+    "cluster.x-k8s.io/wasmtime-spin-v1": "true"
+```
+
+In the preceding yaml, note the nodeSelector and the label. The Kubernetes scheduler will select nodes with the `cluster.x-k8s.io/wasmtime-slight-v1: "true"` or the `cluster.x-k8s.io/wasmtime-spin-v1: "true"` to determine where to schedule Wasm workloads.
```
Comment (mboersma) [inline docs/book/src/topics/wasi.md:109]:
```suggestion
In the preceding YAML, note the nodeSelector and the label. The Kubernetes scheduler will select nodes with the `cluster.x-k8s.io/wasmtime-slight-v1: "true"` or the `cluster.x-k8s.io/wasmtime-spin-v1: "true"` to determine where to schedule Wasm workloads.
```

**Review by mboersma — COMMENTED** [review]
This is great writing: very clear and task-focused. I had some small, mostly optional comments.

File: docs/book/src/topics/wasi.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+## Slight (SpiderLightning)
+Slight (or Spiderlightning) is an open source wasmtime-based runtime that provide cloud capabilities to Wasm microservices. These capabilities include key/value, pub/sub, and much more.
+
+## Fermyon Spin
+"Spin is an open source framework for building and running fast, secure, and composable cloud microservices with WebAssembly. It aims to be the easiest way to get started with WebAssembly microservices, and takes advantage of the latest developments in the WebAssembly component model and Wasmtime runtime."
+
+### Applying the Wasm Runtime Classes
+By default, CAPZ virtual machines include containerd shims to run both `slight` and `spin` workloads. To inform Kubernetes about the ability to run these workloads on CAPZ nodes, you will need to apply a runtime class for each runtime.
```
Comment (CecileRobertMichon) [inline docs/book/src/topics/wasi.md:16]:
```suggestion
By default, CAPZ reference virtual machine images include containerd shims to run both `slight` and `spin` workloads. To inform Kubernetes about the ability to run these workloads on CAPZ nodes, you will need to apply a runtime class for each runtime.
```

Comment (CecileRobertMichon) [inline docs/book/src/topics/wasi.md:16]: > you will need to apply a runtime class for each runtime

I'm assuming we're talking about applying the manifest on the workload cluster here? could be more explicit

Comment (CecileRobertMichon) [reply docs/book/src/topics/wasi.md:16]: Actually, is there a minimum image-builder / k8s version for which our images have this? My first attempt failed with

> Failed to create pod sandbox: rpc error: code = Unknown desc = failed to get sandbox runtime: no runtime for "spin" is configured

I'm using k8s v1.26.0

Comment (CecileRobertMichon) [reply docs/book/src/topics/wasi.md:16]: looks like https://github.com/kubernetes-sigs/image-builder/pull/1037 went in Dec 22nd and v1.26.0 was published Dec 12th so I might need a newer image (I wouldn't expect a user to know how to find this out easily so worth documenting IMO)

Comment (devigned) [reply docs/book/src/topics/wasi.md:16]: I believe I call out the K8s versions in the overview section. Is that sufficient?

Comment (devigned) [reply docs/book/src/topics/wasi.md:16]: Per your error, that is expected given there is no runtime configured for spin in the containerd config.

**Comment by devigned** [comment]
Thank you for the fantastic feedback, @mboersma, @CecileRobertMichon, and @Mossaka. The suggested changes helped a ton.

I have addressed each of them. I believe this is ready for another round of reviews when you have a minute.

File: docs/book/src/topics/wasi.md (hunk 7 above)
Comment (mboersma) [reply docs/book/src/topics/wasi.md]: I think only the Azure ref images for v1.26.1 and later have the WASM shims included.

Comment (devigned) [reply docs/book/src/topics/wasi.md]: @mboersma why would they not have been included in say v1.23.16 and others that were published at the same time as 1.26.1?

Comment (mboersma) [reply docs/book/src/topics/wasi.md]: @devigned, sorry, I misspoke. I was thinking for Cecile that she needs v1.26.1, but all the recent set of patches (v1.23.16 and friends) have the Wasm shims.

Comment (CecileRobertMichon) [reply docs/book/src/topics/wasi.md]: > I believe I call out the K8s versions in the overview section. Is that sufficient?

Ah I somehow missed that (or rather read it then forgot about it by the time I got the next step 😬 . Should be good enough!

File: docs/book/src/topics/wasi.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  type: LoadBalancer
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 80
+  selector:
+    app: wasm-spin
+```
+
+The preceding deployment and service will create a load-balanced "hello world" service with 3 Spin microservices. Note the `runtimeClassName` applied to the Deployment, `wasmtime-spin-v1`, which informs containerd on the cluster node to run the workload with the spin shim.
```
[the remark this answers is not in the document]
Comment (devigned) [reply docs/book/src/topics/wasi.md:81]: I added a section to describe how to curl the services and how to get started building `slight` and `spin` microservices.

Thank you again for pointing out that the URI was not obvious.

File: docs/book/src/topics/wasi.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+### A Running Spin Microservice
+With the service and the deployment applied, you should now have a running Spin microservice running in your workload cluster. If you run the following command against the workload cluster, you can find the IP for the `wasm-spin` service.
```
Comment (mboersma) [inline docs/book/src/topics/wasi.md:84]:
```suggestion
With the service and the deployment applied, you should now have a Spin microservice running in your workload cluster. If you run the following command against the workload cluster, you can find the IP for the `wasm-spin` service.
```

File: docs/book/src/topics/wasi.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+```shell
+curl http://20.121.244.48/hello
+Hello world from Spin!
+```
+
+In the preceding output, we see the HTTP response from our Spin microservice, "Hello world from Spin!".
+
+### Building a Spin or Slight Application
+At this point, you might be asking how do I build my own Wasm microservice. Here are a couple pointers to help you get started.
```
Comment (mboersma) [inline docs/book/src/topics/wasi.md:105]:
```suggestion
At this point, you might be asking "How do I build my own Wasm microservice?" Here are a couple pointers to help you get started.
```

File: docs/book/src/topics/wasi.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (mboersma) [reply docs/book/src/topics/wasi.md:81]: Thanks for adding the new section! I followed the instructions robotically and succeeded:

```
Hello world from Spin!
```

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm

**Comment by devigned** [comment]
@mboersma you are my favorite editor :). You are awesome, my friend!

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

**Pull request merged**